algebird-core/src/main/scala/com/twitter/algebird/macros/Cuber.scala

```diff
@@ -0,0 +1,81 @@
+package com.twitter.algebird.macros
+
+import scala.language.experimental.{ macros => sMacros }
+import scala.reflect.macros.Context
+import scala.reflect.runtime.universe._
+
+// import com.twitter.algebird._
+
+trait Cuber[I] {
+  type K
+  def apply(in: I): TraversableOnce[K]
+}
+
+trait Roller[I] {
+  type K
+  def apply(in: I): TraversableOnce[K]
+}
+
+object Cuber {
+  def cuber[T]: Cuber[T] = macro cuberImpl[T]
```

Comment:
can we add a method like:

```scala
  def cube[K, V](t: TraversableOnce[(K, V)])(implicit c: Cuber[T], sg: Semigroup[V]): Map[c.K, V] =
    MapAlgebra.sumByKey(t.toIterator.flatMap { case (k, v) => c(k).map((_, v)) })
```
and similarly for rollup?

(and tests? kthxbai)